flocker/node/agents/ebs.py

```diff
@@ -390,103 +427,64 @@ def ec2_client(region, zone, access_key_id, secret_access_key):
-        wrapped object can be found.
+    :param func method_name: The method to call.
 
-    :return: A function which will call the method of the wrapped object and do
+    :return: A function which will call the method and do
         the extra exception logging.
     """
-    def _run_with_logging(self, *args, **kwargs):
+    def _run_with_logging(*args, **kwargs):
         """
-        Run given boto.ec2.connection.EC2Connection method with exception
-        logging for ``EC2ResponseError``.
+        Run given boto3.ec2.ServiceResource method with exception
+        logging for ``ClientError``.
         """
-        original = getattr(self, original_name)
-        method = getattr(original, method_name)
-
-        # Trace IBlockDeviceAPI ``method`` as Eliot Action.
-        # See https://clusterhq.atlassian.net/browse/FLOC-2054
-        # for ensuring all method arguments are serializable.
-        with AWS_ACTION(operation=[method_name, args, kwargs]):
-            try:
-                return method(*args, **kwargs)
-            except EC2ResponseError as e:
-                BOTO_EC2RESPONSE_ERROR(
-                    aws_code=e.code,
-                    aws_message=e.message,
-                    aws_request_id=e.request_id,
-                ).write()
-                raise
+        with AWS_ACTION(operation=[method.__name__, args[1:], kwargs]):
+            return method(*args, **kwargs)
     return _run_with_logging
 
 
-def boto_logger(*args, **kwargs):
+def _get_volume_tag(volume, name):
     """
-    Decorator to log all callable boto.ec2.connection.EC2Connection
-    methods.
+    Retrieve the tag from the specified volume with the specified name.
 
-    :return: A function that will decorate all methods of the given
-        class with Boto exception logging.
-    """
-    def _class_decorator(cls):
-        for attr in EC2Connection.__dict__:
-            # Log wrap all callable methods except `__init__`.
-            if attr != '__init__':
-                attribute = getattr(EC2Connection, attr)
-                if callable(attribute):
-                    setattr(cls, attr,
-                            _boto_logged_method(attr, *args, **kwargs))
-        return cls
-    return _class_decorator
-
-
-@boto_logger("connection")
-class _LoggedBotoConnection(PClass):
-    """
-    Wrapper ``PClass`` around ``boto.ec2.connection.EC2Connection``
-    to facilitate logging of exceptions from Boto APIs.
+    :param Volume volume: The volume.
+    :param unicode name: The tag name.
 
-    :ivar boto.ec2.connection.EC2Connection connection: Object
-        representing connection to an EC2 instance with logged
-        ``EC2ConnectionError``.
+    :return: A ``str`` representing the value of the tag, or ``None`` if
+        no tag with this name can be found in the volume.
     """
-    connection = field(mandatory=True)
+    for tag in volume.tags:
+        if tag['Key'] == name:
+            return tag['Value']
+    return None
```

Comment:
It's better to raise an exception rather than return `None`. Probably custom `NoSuchTag` exception. That way if e.g. there's a bug in the code or unexpected edge case where tag is missing then you get a meaningful exception instead of getting (five lines later) a TypeError about how `UUID` can't deal with `None` or errors like that.